src/component/2d/Contours.tsx

```diff
@@ -1,115 +1,169 @@
 }
 
+interface ContoursInnerProps {
+  data: Array<Datum2D>;
+  displayerKey: string;
+}
+
+function usePath(datum: Datum2D, contours) {
+  const { margin, width, height, xDomain, yDomain } = useChartData();
+  const { xShift, yShift } = getShift(datum);
+
+  const _scaleX = get2DXScale({ margin, width, xDomain });
+  const _scaleY = get2DYScale({ margin, height, yDomain });
+  const pathBuilder = new PathBuilder();
+  for (const element of contours) {
+    if (element.lines) {
+      const lines = element.lines;
+      if (lines.length < 1e6) {
+        for (let i = 0; i < lines.length; i += 4) {
+          pathBuilder.moveTo(
+            _scaleX(lines[i] + xShift),
+            _scaleY(lines[i + 1] + yShift),
+          );
+          pathBuilder.lineTo(
+            _scaleX(lines[i + 2] + xShift),
+            _scaleY(lines[i + 3] + yShift),
+          );
+        }
+      }
+    } else {
+      pathBuilder.moveTo(
+        _scaleX(element[0].x + xShift),
+        _scaleY(element[0].y + yShift),
+      );
+
+      for (let j = 1; j < element.length; j++) {
+        pathBuilder.lineTo(
+          _scaleX(element[j].x + xShift),
+          _scaleY(element[j].y + yShift),
+        );
+      }
+    }
+  }
+
+  pathBuilder.closePath();
+
+  return pathBuilder.toString();
+}
+
+const useContoursLevel = (datum: Datum2D, sign: LevelSign) => {
+  const {
+    view: {
+      zoom: { levels },
+    },
+  } = useChartData();
+  const {
+    id,
+    display: { contourOptions },
+  } = datum;
+  const defaultLevel = getDefaultContoursLevel(contourOptions);
+  const level = levels?.[id]?.[sign];
+  return isNaN(level) ? defaultLevel[sign] : level;
+};
+
 function ContoursPaths({
   id: spectrumID,
   sign,
   color,
   datum,
+  onTimeout,
 }: ContoursPathsProps) {
-  const { margin, width, height, xDomain, yDomain, contours } = useChartData();
-
+  const pathRef = useRef<SVGPathElement>(null);
   const activeSpectrum = useActiveSpectrum();
   const preferences = usePreferences();
-  const { xShift, yShift } = getShift(datum);
 
-  const isActive = useMemo(() => {
-    return activeSpectrum === null ? true : spectrumID === activeSpectrum.id;
-  }, [activeSpectrum, spectrumID]);
-
-  function buildContourPath(data) {
-    const _scaleX = get2DXScale({ margin, width, xDomain });
-    const _scaleY = get2DYScale({ margin, height, yDomain });
-    const pathBuilder = new PathBuilder();
-    for (const element of data) {
-      if (element.lines) {
-        const lines = element.lines;
-        if (lines.length < 1e6) {
-          for (let i = 0; i < lines.length; i += 4) {
-            pathBuilder.moveTo(
-              _scaleX(lines[i] + xShift),
-              _scaleY(lines[i + 1] + yShift),
-            );
-            pathBuilder.lineTo(
-              _scaleX(lines[i + 2] + xShift),
-              _scaleY(lines[i + 3] + yShift),
-            );
-          }
-        }
+  const level = useContoursLevel(datum, sign);
+
+  useLayoutEffect(() => {
```

Comment:
Why do you use a layout effect to set the opacity?

Reply:
No for a specific reason and it does not matter if I use useLayoutEffect or useEffect in this case.

Reply:
My question is why do you use an effect at all

Reply:
In the previous version of my code, I use mem hook for the whole path component which was wrong because in any way the component will need to be rendered if we zoom in/out or change the level of the contours so then I memorized the contours calculation to prevent the contours re-calculation when zooming in and out and forgeted to remove the ref and the useEffect that i use for change the color of the contours.

 Thank you for spotting the light on this part of the code, I will remove it and fix the code